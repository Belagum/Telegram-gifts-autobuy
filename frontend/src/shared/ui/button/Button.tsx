// SPDX-License-Identifier: Apache-2.0
// Copyright 2025 Vova Orig

import type { ButtonHTMLAttributes, ReactNode } from "react";
import React from "react";
import clsx from "clsx";
import "./button.css";

export type ButtonVariant = "primary" | "secondary" | "danger" | "ghost";
export type ButtonSize = "sm" | "md" | "lg";

export interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  variant?: ButtonVariant;
  size?: ButtonSize;
  loading?: boolean;
  icon?: ReactNode;
}

export const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  ({ variant = "primary", size = "md", loading = false, disabled, className, children, icon, ...rest }, ref) => {
    const isDisabled = disabled || loading;
    return (
      <button
        ref={ref}
        className={clsx("ui-button", `ui-button--${variant}`, `ui-button--${size}`, className, {
          "ui-button--loading": loading,
        })}
        disabled={isDisabled}
        {...rest}
      >
        {icon && <span className="ui-button__icon">{icon}</span>}
        <span className="ui-button__label">{children}</span>
      </button>
    );
  },
);

Button.displayName = "Button";
