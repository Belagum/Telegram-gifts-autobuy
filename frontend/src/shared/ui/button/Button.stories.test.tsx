// SPDX-License-Identifier: Apache-2.0
// Copyright 2025 Vova Orig

import { render } from "@testing-library/react";
import { Button } from "./Button";
import * as stories from "./Button.stories";

describe("Button stories", () => {
  it("renders primary button", () => {
    const { container } = render(<Button {...(stories.Primary.args ?? {})}>Кнопка</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders loading state", () => {
    const { container } = render(
      <Button {...(stories.Loading.args ?? {})}>Загрузка</Button>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
