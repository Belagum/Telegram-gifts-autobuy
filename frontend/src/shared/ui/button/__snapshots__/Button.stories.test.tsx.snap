// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button stories > renders loading state 1`] = `
<button
  class="ui-button ui-button--primary ui-button--md ui-button--loading"
  disabled=""
>
  <span
    class="ui-button__label"
  >
    Загрузка
  </span>
</button>
`;

exports[`Button stories > renders primary button 1`] = `
<button
  class="ui-button ui-button--primary ui-button--md"
>
  <span
    class="ui-button__label"
  >
    Кнопка
  </span>
</button>
`;
